### PR TITLE
Update log

### PR DESCRIPTION
**Pull request by lplewa**
Use logger in umf. This change also extends logger with the alternative for perror function. 

### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [ ] CI workflows execute properly

- [ ] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [ ] New tests added, especially if they will fail without my changes
- [ ] Added/extended example(s) to cover this functionality
- [ ] Extended the README/documentation
- [ ] All newly added source files have a license
- [ ] All newly added source files are referenced in CMake files
- [x] Logger (with debug/info/... messages) is used

### CODE REVIEW

File: benchmark/ubench.c [hunk 1]
```diff
@@ -44,7 +44,6 @@
 #define DISJOINT_POOL_MAX_POOLABLE_SIZE (2 * ALLOC_SIZE)
```
Comment (bratpiorka) [inline benchmark/ubench.c:44]: bump dates please here are in other files :)

File: src/provider/provider_os_memory.c [hunk 2]
```diff
@@ -308,39 +291,42 @@ static void os_finalize(void *provider) {
 static umf_result_t os_get_min_page_size(void *provider, void *ptr,
                                          size_t *page_size);
 
+#ifndef NDEBUG
 static void print_numa_nodes(os_memory_provider_t *os_provider, void *addr,
                              size_t size) {
 
+    // This function do not use logger as it is expected to print a message
+    // to stdout/stderr
```
Comment (bratpiorka) [inline src/provider/provider_os_memory.c:299]: do not use logger? I see LOG_DEBUG few next lines below

File: src/provider/provider_os_memory.c [hunk 3]
```diff
@@ -308,39 +291,42 @@ static void os_finalize(void *provider) {
 static umf_result_t os_get_min_page_size(void *provider, void *ptr,
                                          size_t *page_size);
 
+#ifndef NDEBUG
```
Comment (bratpiorka) [inline src/provider/provider_os_memory.c:294]: why #ifndef NDEBUG?

File: src/provider/provider_tracking.c [hunk 4]
```diff
@@ -12,6 +12,7 @@
 #include "critnib.h"
 #include "utils_common.h"
 #include "utils_concurrency.h"
+#include "utils_log.h"
 
 #include <umf/memory_pool.h>
 #include <umf/memory_provider.h>
```
Comment (bratpiorka) [inline src/provider/provider_tracking.c:18]: add log to line 58

File: src/provider/provider_tracking.c [hunk 5]
```diff
@@ -84,12 +85,12 @@ umf_memory_pool_handle_t umfMemoryTrackerGetPool(const void *ptr) {
...
-        fprintf(stderr, "tracker is not created\n");
+        LOG_ERR("tracker is not created");
         return NULL;
     }
 
     if (TRACKER->map == NULL) {
-        fprintf(stderr, "tracker's map is not created\n");
+        LOG_ERR("tracker's map is not created");
         return NULL;
     }
```
Comment (bratpiorka) [inline src/provider/provider_tracking.c:96]: add log to line 135

File: src/utils/utils_log.c [hunk 6]
```diff
@@ -31,6 +32,7 @@
 
 #define LOG_MAX 8192
 #define LOG_HEADER 256
+#define LOG_PERROR 1024
```
Comment (bratpiorka) [inline src/utils/utils_log.c:35]: where this is used?

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: src/base_alloc/base_alloc_global.c [hunk 7]
```diff
@@ -163,18 +163,16 @@ void *umf_ba_global_aligned_alloc(size_t size, size_t alignment) {
 
     int ac_index = size_to_idx(size);
     if (ac_index >= NUM_ALLOCATION_CLASSES) {
-        // TODO: use logger here
-        // fprintf(stderr,
-        //         "base_alloc: allocation size (%zu) larger than the biggest "
-        //         "allocation class. Falling back to OS memory allocation.\n",
-        //         size);
+        LOG_ERR("base_alloc: allocation size (%zu) larger than the biggest "
```
Comment (igchor) [inline src/base_alloc/base_alloc_global.c:166]: This should be warning I think

File: src/base_alloc/base_alloc_global.c [hunk 8]
```diff
@@ -163,18 +163,16 @@ void *umf_ba_global_aligned_alloc(size_t size, size_t alignment) {
...
+                "allocation class. Falling back to OS memory allocation.",
+                size);
         return add_metadata_and_align(ba_os_alloc(size), size, alignment);
     }
 
     if (!BASE_ALLOC.ac[ac_index]) {
         // if creating ac failed, fall back to os allocation
-        fprintf(stderr, "base_alloc: allocation class not created. Falling "
-                        "back to OS memory allocation.\n");
+        LOG_ERR("base_alloc: allocation class not created. Falling "
```
Comment (igchor) [inline src/base_alloc/base_alloc_global.c:174]: This should be warning I think

File: src/provider/provider_os_memory.c [hunk 9]
```diff
@@ -422,12 +408,10 @@ static umf_result_t os_alloc(void *provider, size_t size, size_t alignment,
     }
 
     if (alignment && (alignment % page_size) && (page_size % alignment)) {
-        if (os_provider->traces) {
-            fprintf(stderr,
-                    "wrong alignment: %zu (not a multiple or a divider of the "
-                    "minimum page size (%zu))\n",
-                    alignment, page_size);
-        }
+        LOG_DEBUG("wrong alignment: %zu (not a multiple or a divider of the "
```
Comment (igchor) [inline src/provider/provider_os_memory.c:411]: LOG_ERROR? we are returning an error here after all

File: src/provider/provider_os_memory.c [hunk 10]
```diff
@@ -486,11 +464,11 @@ static umf_result_t os_alloc(void *provider, size_t size, size_t alignment,
...
     }
 
-    if (os_provider->traces) {
-        // TODO: if we don't touch the page, we'll get EFAULT from move_pages.
-        // Should we add an option to touch pages after the allocation?
-        print_numa_nodes(os_provider, addr, size);
-    }
+#ifndef NDEBUG
+    // TODO: if we don't touch the page, we'll get EFAULT from move_pages.
+    // Should we add an option to touch pages after the allocation?
```
Comment (igchor) [inline src/provider/provider_os_memory.c:469]: Yes, we probably should.

File: src/utils/utils_log.c [hunk 11]
```diff
@@ -82,17 +85,73 @@ void util_log(util_log_level_t level, const char *format, ...) {
...
+
+    int tmp = vsnprintf(buffer, sizeof(buffer), format, args);
+    ASSERT(tmp > 0);
+
+    b_pos += (int)tmp;
+    b_size -= (int)tmp;
+
+    const char *postfix = "";
+
+    if (perror) {
```
Comment (igchor) [inline src/utils/utils_log.c:99]: nit: could you move this to a separate function?

File: benchmark/ubench.c (hunk 1 above)
Comment (lplewa) [reply benchmark/ubench.c:44]: Done.

File: src/provider/provider_os_memory.c (hunk 2 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done.

File: src/provider/provider_os_memory.c (hunk 3 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done.

File: src/provider/provider_tracking.c (hunk 4 above)
Comment (lplewa) [reply src/provider/provider_tracking.c:18]: Done.

File: src/provider/provider_tracking.c (hunk 5 above)
Comment (lplewa) [reply src/provider/provider_tracking.c:94]: Done.

File: src/utils/utils_log.c (hunk 6 above)
Comment (lplewa) [reply src/utils/utils_log.c]: Done.

File: src/provider/provider_os_memory.c (hunk 10 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: I removed this code.

File: src/provider/provider_os_memory.c (hunk 9 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done.

File: src/base_alloc/base_alloc_global.c (hunk 8 above)
Comment (lplewa) [reply src/base_alloc/base_alloc_global.c]: Done.

File: src/base_alloc/base_alloc_global.c (hunk 7 above)
Comment (lplewa) [reply src/base_alloc/base_alloc_global.c]: Done.

File: benchmark/ubench.c (hunk 1 above)
Comment (lukaszstolarczuk) [reply benchmark/ubench.c:44]: not done?

File: src/provider/provider_os_memory.c [hunk 12]
```diff
@@ -308,39 +291,41 @@ static void os_finalize(void *provider) {
...
         if (hwloc_bitmap_list_snprintf(os_provider->nodeset_str_buf,
                                        NODESET_STR_BUF_LEN, nodeset)) {
-            printf("alloc(%zu) = 0x%llx, allocate on NUMA nodes = %s\n", size,
-                   (unsigned long long)addr, os_provider->nodeset_str_buf);
+            LOG_DEBUG("alloc(%zu) = 0x%llx, allocate on NUMA nodes = %s\n",
+                      size, (unsigned long long)addr,
+                      os_provider->nodeset_str_buf);
         } else {
-            fprintf(stderr, "cannot print assigned NUMA node\n");
+            LOG_DEBUG("cannot print assigned NUMA node\n");
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory.c:322]: drop `\n` I guess... (here and above)

File: src/base_alloc/base_alloc.c [hunk 13]
```diff
@@ -288,7 +289,7 @@ void umf_ba_destroy(umf_ba_pool_t *pool) {
 #ifndef NDEBUG
     ba_debug_checks(pool);
     if (pool->metadata.n_allocs) {
-        fprintf(stderr, "umf_ba_destroy(): pool->metadata.n_allocs = %zu\n",
+        LOG_ERR("umf_ba_destroy(): pool->metadata.n_allocs = %zu\n",
```
Comment (lukaszstolarczuk) [inline src/base_alloc/base_alloc.c:292]: drop `\n` I guess

File: src/provider/provider_os_memory.c (hunk 12 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/base_alloc/base_alloc.c (hunk 13 above)
Comment (lplewa) [reply src/base_alloc/base_alloc.c]: Done

File: src/utils/utils_log.h [hunk 14]
```diff
@@ -14,19 +14,38 @@
 extern "C" {
 #endif
 
-typedef enum { LOG_DEBUG, LOG_INFO, LOG_WARNING, LOG_ERROR } util_log_level_t;
+typedef enum {
+    LOG_DEBUG,
+    LOG_INFO,
+    LOG_WARNING,
+    LOG_ERROR,
+    LOG_FATAL
```
Comment (lukaszstolarczuk) [inline src/utils/utils_log.h:22]: do we have any tests for FATAL?

File: benchmark/ubench.c (hunk 1 above)
Comment (lplewa) [reply benchmark/ubench.c:44]: Done?

File: src/utils/utils_log.c (hunk 11 above)
Comment (lplewa) [reply src/utils/utils_log.c:100]: Technically we can, but imho it's not wroth. This code is quite integrated with rest of the function, so it is a bit of extra code to abstract it, especial that i do not expect that we would need as a function to use in different case

**Review by igchor — APPROVED** [review]

**Review by lukaszstolarczuk — COMMENTED** [review]
LGTM, please rebase

**Comment by lplewa** [comment]
> LGTM, please rebase

done

**Pull request merged**